### DISCUSSION TITLE
Autopause fails (eth0 not found, but exists)

### DISCUSSION THREAD

**Discussion by marcso-bdrck** [itzg/docker-minecraft-server#1943] [Q&A]
Autopause fails because eth0 is not found in /sys/class/net/ (only the host side vethXXXX):

```
[init] Autopause functionality enabled
[Autopause loop] Selected interface "eth0" does not exist!
[Autopause loop] Available interfaces within the docker container:
[Autopause loop]   bonding_masters lo vethe5e4612
```

The problem is that ifconfig shows eth0.
I'm not a Docker expert, but something seems to have changed with the handling of sysfs. I set up Docker on Ubuntu 22.04 yesterday. Every container shows the same behaviour.

Any ideas?

**Comment by itzg** [comment]
Providing your container config will probably help for people to answer.

**Reply by marcso-bdrck** [reply]
I have now replicated everything on my home "server" (same OS and docker version). Everything is working as expected here: eth0 can be found in /sys/class/net/. 
The container configuration does not differ from the main servers config (= hosted vserver). 

A reinstallation of the vserver shows the same behaviour as before: no eth0; only vethXXXX in the container. Maybe that's what the hoster (strato.de) intended?

Maybe there's a workaround for using autopause with vethXXXX? (unfortunately vethXXXX ist dynamic)

I would also be very grateful for a good ressource on the topic of sysfs in docker containers.

Edit: 
The problem persits with ervery container, even the minimal example:
docker run -d -it -p 25565:25565 --name mc -e EULA=TRUE itzg/minecraft-server

**Reply by itzg** [reply]
Something is unusual with your Docker installation. Containers encapsulate the network stack, so you shouldn't see the same interfaces within the container as present at the host level.

sysfs in this case is the low level kernel space that tools like ifconfig access.

**Reply by marcso-bdrck** [reply]
The strange thing is that I don't see the same interfaces as on the host level, only the vitual interfaces belonging to the container. Here is a screenshot:

![image](https://user-images.githubusercontent.com/89865143/216547979-6317490a-8b33-4a08-9b65-30e44f1baa22.png)

Docker was installed with apt from the official respo like described at https://docs.docker.com/engine/install/ubuntu/.

**Reply by itzg** [reply]
Please provide the output of:

```
docker run -it --rm --entrypoint ifconfig itzg/minecraft-server

docker run -it --rm --entrypoint ls itzg/minecraft-server /sys/class/net
```

**Reply by marcso-bdrck** [reply]
root@XXXXXXX:~# docker run -it --rm --entrypoint ifconfig itzg/minecraft-server
[..]
e5c51f51dbda: Pull complete
Digest: sha256:6c6a006894ebb4cc0ea91fe8a7e3ded6c8f845909616734f90bbe8d504ae5e69
Status: Downloaded newer image for itzg/minecraft-server:latest
eth0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 172.17.0.5  netmask 255.255.0.0  broadcast 172.17.255.255
        ether 02:42:ac:11:00:05  txqueuelen 0  (Ethernet)
        RX packets 2  bytes 180 (180.0 B)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 0  bytes 0 (0.0 B)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

lo: flags=73<UP,LOOPBACK,RUNNING>  mtu 65536
        inet 127.0.0.1  netmask 255.0.0.0
        loop  txqueuelen 1000  (Local Loopback)
        RX packets 0  bytes 0 (0.0 B)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 0  bytes 0 (0.0 B)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

root@XXXXXXX:~# docker run -it --rm --entrypoint ls itzg/minecraft-server /sys/class/net
bonding_masters  lo  vethe109d1d

**Reply by itzg** [reply]
@Oekn5w , can you help look at this? ifconfig and /sys/class/net contents don't match up in this case. We might need to move away from using that.

**Reply by marcso-bdrck** [reply]
One more small addition:
Container and host interfaces have different labels.
f.e.
/sys/class/net/vethce76850 in container is linked to veth725a22f on the host

**Reply by itzg** [reply]
Next image switches to just using `ifconfig`

https://github.com/itzg/docker-minecraft-server/actions/runs/4092545295

**Reply by Oekn5w** [reply]
uh oh, when you don't remember what you've come up with back then and not find the (likely) Stackoverflow thread to blame ...
Sorry I missed the tag earlier.
Knockd doesn't seem to specify where it gets its interface from, so assuming the change you've made works for the author, I'd say we're good as this is likely the first report in the 3 years since.

**Reply by marcso-bdrck** [reply]
Yes I can imagine, it's a problem that rarely occurs. I think I've been able to narrow down the issue: My hosting service uses Virtuozzo for the vservers. There seems to be some kind of incompatibility causing the vethXXXX label in the container's sys-folder.

But nevermind, thanks for fixing my issue! :)

**Reply by itzg** [reply]
Nothing like a good edge case to shake out all the bugs 😉

**Reply by marcso-bdrck** [reply]
@itzg I'm very sorry but I've found a "bug" in the new implementation:

I think alpine's ifconfig doesn't know argument -s. So ifconfig -s | tail +2 | cut -f1 -d ' ' breaks down f.e. for image itzg/minecraft-server:java8.

```
root@XXXX:~# docker run -it --rm --entrypoint ifconfig itzg/minecraft-server -s
Iface      MTU    RX-OK RX-ERR RX-DRP RX-OVR    TX-OK TX-ERR TX-DRP TX-OVR Flg
eth0      1500        2      0      0 0             0      0      0      0 BMRU
lo       65536        0      0      0 0             0      0      0      0 LRU
root@XXXX:~# docker run -it --rm --entrypoint ifconfig itzg/minecraft-server:java8 -s
ifconfig: -s: error fetching interface information: Device not found
```

**Reply by itzg** [reply]
Sorry about that. I'll rework it. 

Honesty the alpine images are effectively deprecated since they have too many non-compliant behaviors...like what you found.

**Reply by marcso-bdrck** [reply]
I completely overlooked that there are other java8 images. I just use java8-multiarch on Ubuntu. Thank you for everything!
